sys/kern/mcount.c

```diff
@@ -0,0 +1,195 @@
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the University nor the names of its contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include <sys/types.h>
+#include <sys/gmon.h>
+#include <sys/interrupt.h>
+#include <sys/spinlock.h>
+#include <sys/types.h>
+
+static SPIN_DEFINE(mcount_lock, 0);
+
+/*
+ * The function is updating an array of linked lists, which stores 
+ * how many times a functions have been called by another function.
+ * 
+ *  froms[X] - an array of linked list, index X encodes a function, the kept
+ *             value is an index of tos entry (the first node in the list,
+ *             0 - means an empty list), the list is sorted by most recently
+ *             used function
+ * 
+ *  tos[N]   - a node, which let us know how many times function with address selfpc 
+ *             have been called by X, and stores index of the next element in the list
+ *             (0 - means the end of the list)
+ * 
+ *  tos[0] + 1  is the smallest index of an unused tos entry
+ * 
+ * \warning This function while unlocking spinlock can turn on interrupts when 
+ * thread's td_idnest == 0, this problem occurs e.g. in mips_exc_handler,
+ * this is why we do not profile user_mode_p.
+ */
+
+__no_instrument_kgprof void __cyg_profile_func_enter(void *self, void *from) {
+  u_long frompc = (u_long)from, selfpc = (u_long)self;
+  u_short *frompcindex;
+  tostruct_t *top, *prevtop;
+  gmonparam_t *p = &_gmonparam;
+  long toindex;
+
+  if (p->state != GMON_PROF_ON)
+    return;
+
+  WITH_SPIN_LOCK (&mcount_lock) {
+    p->state = GMON_PROF_BUSY;
```

Comment:
As I understand, You don't need that state now. Please add comment that this will be used in near future.

Reply:
It is used in kgmon, cause at the end of the function we can move one node at the beginning and if we call kgmon at that moment we would omit this node.

Reply:
Okay, I see it. Please add comment about it because it is not obvious. (Here and in python code too.)

Reply:
Done!